### PR TITLE
Removes annoying this "would catch all throwables" warnings

### PR DESCRIPTION
**Pull request by odersky**
Silenced all "this would catch all throwables" messages in tools build except one. The one that still warns needs to be fixed by tightening the caught exception to be something different from Throwable. Review by @VladUreche.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala [hunk 1]
```diff
@@ -242,7 +242,7 @@ trait ModelFactoryImplicitSupport {
 
           available = Some(search.tree != EmptyTree)
         } catch {
-          case _ =>
+          case _ => // !!! [Martin]: One should never silently catch all Throwables. Please sharpen the exception type.
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala:245]: It should be case _: TypeError =>. Should I do another pull request with this?

**Comment by VladUreche** [comment]
I can add the TypeError in ModelFactoryImplicitSupport.scala in the next pull request on scaladoc. 
Therefore, LGTM.

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala (hunk 1 above)
Comment (odersky) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala:245]: On Fri, Jul 20, 2012 at 4:43 PM, Vlad Ureche
reply@reply.github.com
wrote:

> > @@ -242,7 +242,7 @@ trait ModelFactoryImplicitSupport {
> > 
> > ```
> >        available = Some(search.tree != EmptyTree)
> >      } catch {
> > ```
> > -          case _ =>
> > -          case _ => // !!! [Martin]: One should never silently catch all Throwables. Please sharpen the exception type.
> 
> It should be case _: TypeError =>. Should I do another pull request with this?

Yes, please do! Thanks - Martin

> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/955/files#r1205800

## 

Martin Odersky
Prof., EPFL and Chairman, Typesafe
PSED, 1015 Lausanne, Switzerland
Tel. EPFL: +41 21 693 6863
Tel. Typesafe: +41 21 691 4967

File: src/compiler/scala/tools/util/VerifyClass.scala [hunk 2]
```diff
@@ -13,7 +13,7 @@ object VerifyClass {
       Class.forName(name, true, cl)
       (name, None)
     } catch {
-      case x => (name, Some(x.toString))
+      case x: Throwable => (name, Some(x.toString))
```
Comment (jsuereth) [inline src/compiler/scala/tools/util/VerifyClass.scala:16]: Note:  This one is my fault.  We just need something that captures Class VerifyError (and related) exceptions as well as ExceptionInInitializationErrors for static objects that bomb on classload.    Since this class is only used in our build, it's not critical for a release.

**Comment by paulp** [comment]
And I love retronym as much as the next guy (more even) but let's never again accept a pull request which creates a new warning unless it is accompanied by code which deals with the warning in our own codebase.  Any theory that we would all leap out of our chairs to respond to the warnings has been thoroughly debunked.

File: src/compiler/scala/tools/util/VerifyClass.scala (hunk 2 above)
Comment (odersky) [inline src/compiler/scala/tools/util/VerifyClass.scala:16]: On Fri, Jul 20, 2012 at 4:46 PM, Josh Suereth
reply@reply.github.com
wrote:

> > @@ -13,7 +13,7 @@ object VerifyClass {
> >        Class.forName(name, true, cl)
> >        (name, None)
> >      } catch {
> > -      case x => (name, Some(x.toString))
> > -      case x: Throwable => (name, Some(x.toString))
> 
> Note:  This one is my fault.  We just need something that captures Class VerifyError (and related) exceptions as well as ExceptionInInitializationErrors for static objects that bomb on classload.    Since this class is only used in our build, it's not critical for a release.
> 
> my criterion was: as long as there's no obvious control flow exception
> going through the catch and the caught exception is processed further
> I just put in the Throwable.

Cheers
- Martin
  > ---
  > Reply to this email directly or view it on GitHub:
  > https://github.com/scala/scala/pull/955/files#r1205836

## 

Martin Odersky
Prof., EPFL and Chairman, Typesafe
PSED, 1015 Lausanne, Switzerland
Tel. EPFL: +41 21 693 6863
Tel. Typesafe: +41 21 691 4967

**Comment by odersky** [comment]
On Fri, Jul 20, 2012 at 4:50 PM, Paul Phillips
reply@reply.github.com
wrote:

> And I love retronym as much as the next guy (more even) but let's never again accept a pull request which creates a new warning unless it is accompanied by code which deals with the warning in our own codebase.  Any theory that we would all leap out of our chairs to respond to the warnings has been thoroughly debunked.

+1
- Martin

> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/955#issuecomment-7132077

## 

Martin Odersky
Prof., EPFL and Chairman, Typesafe
PSED, 1015 Lausanne, Switzerland
Tel. EPFL: +41 21 693 6863
Tel. Typesafe: +41 21 691 4967

**Comment by adriaanm** [comment]
since this conflicted with the scaladoc commit, I'll rebase it and resubmit it soon.

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala (hunk 1 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala:245]: I'll take care of it together with reviews from other PR, as I need to rebase this one anyway.

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryImplicitSupport.scala:245]: Thanks Adriaan!

**Comment by retronym** [comment]
Sorry about the noisy warnings at an inopportune moment. Hindsight suggests it would have been better for me to add the `_: Throwable` everywhere in the original PR and ask reviewers to spot places that could be loosened. (Which turns out to be not so many, sadly.)

**Comment by paulp** [comment]
It occurred to me we could make this a little more to the point.  There are other reasons not to catch everything of course, but this is the one I see confuse people the most:

```
object Foo {
  def foo(l: List[Int]): Int = {
    try l foreach { _ => return 5 }
    catch { case x => 11 }
    22
  }
}
// 
// ./build/pack/bin/scalac -d /tmp ./a.scala 
// ./a.scala:4: warning: This catches all Throwables. If this is really intended, use `case x : Throwable` to clear this warning.
//         catch { case x => 11 }
//                      ^
// ./a.scala:3: warning: warning: catch block may intercept non-local return from method foo
//         try l foreach { _ => return 5 }
//               ^
// two warnings found
```
